contracts/contracts/ZKGroups.sol

```diff
@@ -0,0 +1,197 @@
+        semaphore = ISemaphore(semaphoreAddress);
+
+        for (uint8 i = 0; i < _verifiers.length; ) {
+            verifiers[_verifiers[i].merkleTreeDepth] = IVerifier(_verifiers[i].contractAddress);
+
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+    /// @dev See {IZKGroups-updateOffchainGroups}.
+    function updateOffchainGroups(OffchainGroup[] calldata _offchainGroups) external override onlyOwner {
+        for (uint8 i = 0; i < _offchainGroups.length; ){
+            uint256 groupId = uint256(keccak256(abi.encodePacked("Offchain_",_offchainGroups[i].name))) % SNARK_SCALAR_FIELD;
+            
+            _updateOffchainGroup(groupId, _offchainGroups[i]);
+
+            unchecked {
+                ++i;
+            }
+        }
+    }
+
+    /// @dev See {IZKGroups-verifyOffchainGroupProof}.
+    function verifyOffchainGroupProof(
+        uint256 groupId,
+        bytes32 signal,
+        uint256 nullifierHash,
+        uint256 externalNullifier,
+        uint256[8] calldata proof
+    ) external override {
+        uint256 merkleTreeRoot = getOffchainRoot(groupId);
+
+        require(merkleTreeRoot != 0, "Offchain group does not exist");
+
+        require(!nullifierHashes[groupId][nullifierHash], "you cannot use the same nullifier twice");
+
+        uint256 merkleTreeDepth = getOffchainDepth(groupId);
+
+        IVerifier verifier = verifiers[merkleTreeDepth];
+
+        _verifyProof(signal, merkleTreeRoot, nullifierHash, externalNullifier, proof, verifier);
+
+        _saveNullifierHash(groupId, nullifierHash);
+
+        emit ProofVerified(groupId, nullifierHash, externalNullifier, signal);
+    }
+
+    /// @dev See {IZKGroups-getOffchainRoot}.
+    function getOffchainRoot(uint256 groupId) public view override returns (uint256) {
+        return offchainGroups[groupId].merkleTreeRoot;
+    }
+
+    /// @dev See {IZKGroups-getOffchainDepth}.
+    function getOffchainDepth(uint256 groupId) public view override returns (uint256) {
+        return offchainGroups[groupId].merkleTreeDepth;
+    }
+
+    /// @dev Updates an off-chain group.
+    /// @param groupId: Id of the group.
+    /// @param offchainGroup: off-chain group data.
+    function _updateOffchainGroup(
+        uint256 groupId,
+        OffchainGroup calldata offchainGroup
+    ) private onlySupportedMerkleTreeDepth(offchainGroup.merkleTreeDepth){
+        offchainGroups[groupId] = offchainGroup;
+
+        emit OffchainGroupUpdated(groupId, offchainGroup.name, offchainGroup.merkleTreeRoot, offchainGroup.merkleTreeDepth);
+    }
+
+    /// @dev Save nullifier hash to avoid double signaling.
+    /// @param groupId: Id of the group.
+    /// @param nullifierHash: Nullifier hash.
+    function _saveNullifierHash(uint256 groupId, uint256 nullifierHash) internal {
+        nullifierHashes[groupId][nullifierHash] = true;
+    }
+
+    /// @dev See {IZKGroups-createOnchainGroup}.
+    function createOnchainGroup(
```

Comment:
@jdhyun09 The functions below and the `onlyOnchainGroupAdmin` modifier can also be removed from this contract, since they can also be used with the official `Semaphore.sol` contract directly.